app/api/controllers/RequestHandler.scala

```diff
@@ -0,0 +1,205 @@
+import cats.data.EitherT
+import cats.implicits._
+import play.api.http.Status
+import play.api.libs.json.{JsValue, Writes}
+import play.api.mvc.Result
+import play.api.mvc.Results.InternalServerError
+import utils.Logging
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait RequestHandler[InputRaw <: RawData] {
+
+  def handleRequest(rawData: InputRaw)(implicit ctx: RequestContext, request: UserRequest[_], ec: ExecutionContext): Future[Result]
+
+}
+
+object RequestHandler {
+
+  def withParser[InputRaw <: RawData, Input](parser: RequestParser[InputRaw, Input]): ParserOnlyBuilder[InputRaw, Input] =
+    new ParserOnlyBuilder[InputRaw, Input](parser)
+
+  // Intermediate class so that the compiler can separately capture the InputRaw and Input types here, and the Output type later
+  class ParserOnlyBuilder[InputRaw <: RawData, Input] private[RequestHandler](parser: RequestParser[InputRaw, Input]) {
+
+    def withService[Output](
+                             serviceFunction: Input => Future[Either[ErrorWrapper, ResponseWrapper[Output]]]): RequestHandlerBuilder[InputRaw, Input, Output] =
+      RequestHandlerBuilder(parser, serviceFunction)
+
+  }
+
+  case class RequestHandlerBuilder[InputRaw <: RawData, Input, Output] private[RequestHandler](
+                                                                                                parser: RequestParser[InputRaw, Input],
+                                                                                                service: Input => Future[Either[ErrorWrapper, ResponseWrapper[Output]]],
+                                                                                                errorHandling: ErrorHandling = ErrorHandling.Default,
+                                                                                                resultCreator: ResultCreator[InputRaw, Input, Output] = ResultCreator.noContent[InputRaw, Input, Output](),
+                                                                                                auditHandler: Option[AuditHandler] = None
+                                                                                              ) extends RequestHandler[InputRaw] {
+
+    def handleRequest(rawData: InputRaw)(implicit ctx: RequestContext, request: UserRequest[_], ec: ExecutionContext): Future[Result] =
+      Delegate.handleRequest(rawData)
+
+    def withResultCreator(resultCreator: ResultCreator[InputRaw, Input, Output]): RequestHandlerBuilder[InputRaw, Input, Output] =
+      copy(resultCreator = resultCreator)
+
+    def withErrorHandling(errorHandling: ErrorHandling): RequestHandlerBuilder[InputRaw, Input, Output] =
+      copy(errorHandling = errorHandling)
+
+    def withAuditing(auditHandler: AuditHandler): RequestHandlerBuilder[InputRaw, Input, Output] =
+      copy(auditHandler = Some(auditHandler))
+
+    /** Shorthand for
+      * {{{
+      * withResultCreator(ResultCreator.plainJson(successStatus))
```

Comment:
These comments seem redundant, as they are just repeating the code.

Reply:
These comments are part of the common code classes. As discussed, refining or refactoring it can be looked into separately.